ocaml/xapi-guard/lib/lwt_bounded_stream.ml

```diff
@@ -0,0 +1,48 @@
+(*
+ * Copyright (c) 2012 Citrix Systems
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+let ( let* ) = Lwt.bind
+
+type 'a t = {stream: 'a Lwt_stream.t; capacity: int; size: int ref}
+
+let create capacity =
+  let stream, stream_push = Lwt_stream.create () in
+  let t = {stream; capacity; size= ref 0} in
+  let push = function
+    | Some _ when !(t.size) > t.capacity ->
+        None
```

Comment:
Wouldn't this result in data loss? When we're over capacity we're dropping writes, but we do have some elements in the stream already.
I see that the caller would set the queue to disengaged, although for safety should we just cancel/wipe all entries from this queue immediately when this happens? (we'll switch to disengaged mode anyway). Or use some other kind of boolean in this record and raise an exception if we attempt to use it?


Perhaps we should use a different name for this module even if we keep it, when I was reviewing the rest of the code I assumed this is the bounded push from Lwt (with the blocking semantics), not this one with the dropping semantics.

Reply:
The queue elements contain the metadata that identifies the contents in the disk buffer, not the contents themselves. Stopping to push elements to the queue does not lead to data loss.

You are right in that the queue while in disengaged mode is not being emptied. This was in the comments, but was not happening.